pdns/slavecommunicator.cc

```diff
@@ -710,6 +710,19 @@ void CommunicatorClass::addSlaveCheckRequest(const DomainInfo& di, const ComboAd
   Lock l(&d_lock);
   DomainInfo ours = di;
   ours.backend = 0;
+  string remote_address = remote.toString();
+
+  // When adding a check, if the remote addr from which notification was
+  // received is a master, clear all other masters so we can be sure the
+  // query goes to that one.
+  for (const auto& master : ours.masters) {
```

Comment:
I'm not sure we should remove all other masters. If for some reason like a firewall issue we can receive a `NOTIFY` from a master but are unable to open a `TCP` connection to it for the actual `AXFR` we are stuck. Perhaps we should just move this master to the front of the vector so we try it first, but can still fall back to another master if that doesn't work out?

Reply:
@rgacogne the reason why I do it like this is that later on in the code we randomize the list of ours.masters again and ask a random master if there are any fixes, but in the end it'll still go to the first original master.
By removing all the masters temporarily I make sure that it'll only go back to the notifier, which is safer than the current code that always sends it to the same master.
I do agree that if for some weird reason there is only a one way connection, this thing will cause problems, but I also think that the original code will do a worse job in the case you have stated and we should assume that if a target in our masters list does not answer us it's because it's dead and not blocked.
It should be discouraged to have a master in the list that blocks us by firewall.

Reply:
Oh, you are right, I missed the shuffling in `SlaveSenderReceiver::send()`. Perhaps it would make sense to remove the shuffling from `SlaveSenderReceiver::send()`, make sure the notifier is the first masters on the list and shuffle the remaining ones? This way we would try the notifier first, which is your original intent and makes sense to me, but we could still fallback to another one if needed?

Reply:
@rgacogne I can try adding that change, but I think that with how "send" is implemented it's redundant, as "send" shuffles the master list in order to check a random master, as it only works on "masters.begin()", meaning shuffling all masters but the first will not change who we address.

Also, I think we can get to the "send" (from slaveRefresh) through a flow that does not go through a notify (and addSlaveCheckRequest), in which case we do want to shuffle the entire list, and not leave the first master as the first master. We are unware in "send" if we got there through "notify" or not.

Reply:
Oh, you are right, I missed the fact that the `SlaveSenderReceiver::send()` only tries one master even though you did mention it earlier, sorry. In that case your current solution makes sense, although I'm still a bit worried about an asymmetric network issue, but I guess we would need to refactor the code to try more than one master to address it.